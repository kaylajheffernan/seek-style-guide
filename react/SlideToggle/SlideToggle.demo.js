import SlideToggle from './SlideToggle';

export default {
  route: '/slide-toggle',
  title: 'SlideToggle',
  component: SlideToggle,
  initialProps: {
    onToggle: () => {
      console.log('I have been toggled');
    },
    id: 'myToggle',
    label: {
      position: 'right',
      text: 'Toggle me'
    }
  },
  options: [
    {
      label: 'States',
      type: 'checklist',
      states: [
        {
          label: 'Checked',
          transformProps: ({ className, ...props }) => ({
            ...props,
            checked: true
          })
        },
        {
          label: 'Hide label',
          transformProps: ({ className, ...props }) => ({
            ...props,
            label: {
              ...props.label,
              hidden: true
            }
          })
        }
      ]
    },
    {
      label: 'Label position',
      type: 'radio',
      states: [
        {
          label: 'Right',
          transformProps: props => ({
            ...props,
            label: {
              ...props.label,
              position: 'right'
            }
          })
        },
        {
          label: 'Left',
          transformProps: props => ({
            ...props,
            label: {
              ...props.label,
              position: 'left'
            }
          })
        }
      ]
    }
  ]
};
