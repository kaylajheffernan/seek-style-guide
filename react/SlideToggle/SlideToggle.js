import styles from './SlideToggle.less';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import TickIcon from '../TickIcon/TickIcon';
import classnames from 'classnames';

const LEFT = 'left';
const RIGHT = 'right';

export default class SlideToggle extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
    onToggle: PropTypes.func,
    position: PropTypes.oneOf([LEFT, RIGHT]),
    hideLabel: PropTypes.bool,
    children: PropTypes.node,
    checked: PropTypes.bool
  }

  static defaultProps = {
    position: RIGHT,
    hideLabel: false
  };

  handleChange = event => {
    const { onToggle } = this.props;
    if (onToggle) {
      onToggle(event.target.checked);
    }
  }

  renderLabel = () => {
    const { position, text } = this.props;
    return (
      <span
        className={classnames(
          styles.label,
          position === LEFT ? styles.labelLeft : styles.labelRight
        )}>{text}</span>
    );
  }

  render() {
    const { children, id, checked, position, text, hideLabel } = this.props;
    return (
      <label htmlFor={id} className={styles.root}>
        {(hideLabel !== true && position === LEFT && !children) && this.renderLabel()}
        <div className={styles.switch}>
          <input
            type="checkbox"
            id={id}
            aria-label={text}
            className={styles.input}
            checked={checked}
            onChange={this.handleChange}
          />
          <div className={styles.slider}>
            <div className={styles.slideButton}>
              <TickIcon
                className={styles.icon}
                svgClassName={styles.svg}
              />
            </div>
          </div>
        </div>
        {(hideLabel !== true && position === RIGHT && !children) && this.renderLabel()}
      </label>
    );
  }
}
