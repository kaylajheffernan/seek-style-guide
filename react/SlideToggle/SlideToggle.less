@import (reference) "~seek-style-guide/theme";

.root {
  display: inline-flex;
  line-height: @grid-row-height * 3;
  height: @grid-row-height * 3;
  font-size: 14px;
}

.switch {
  display: inline-block;
  position: relative;
  cursor: pointer;
}

.input {
  position: absolute;
  left: 10px;
  top: 5px;
  margin: 0;

  &:focus + .slider {
    box-shadow: 0px 0px 10px 0px @sk-focus;
  }

  &:active + .slider .slideButton {
    transform: scale(1.15, 0.85);
  }

  &:checked + .slider .slideButton {
    background-color: @sk-highlight;
    left: 21px;
  }

  &:checked + .slider .slideButton svg {
    opacity: 1;
  }
}

.label {

}

.labelLeft {
  margin-right: @grid-gutter-width / 2;
}

.labelRight {
  margin-left: @grid-gutter-width / 2;
}

.slider {
  position: relative;
  display: inline-block;
  left: 0;
  width: 48px;
  height: @grid-row-height * 3;
  border-radius: @grid-row-height * 1.5;
  background-color: @sk-mid-gray-light;
  transition: .3s;
}

.slideButton {
  position: absolute;
  left: 0;
  height: @grid-row-height * 3;
  width: @grid-row-height * 3;
  border-radius: 50%;
  background-color: @sk-mid-gray;
  transition: .3s;
  text-align: center;
}

.icon {
  vertical-align: sub;
}

.svg {
  fill: white;
  opacity: 0;
  transition: opacity 0.15s;
}
