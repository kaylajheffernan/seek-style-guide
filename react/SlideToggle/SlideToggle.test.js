import React from 'react';
import renderer from 'react-test-renderer';
import {
  Simulate,
  findRenderedDOMComponentWithTag,
  renderIntoDocument
} from 'react-dom/test-utils';

import SlideToggle from './SlideToggle';

const renderSlideToggle = (props, children) => renderer.create(
  <SlideToggle
    id="testToggle"
    label="Test toggle"
    {...props}>
    {children}
  </SlideToggle>
);

describe('Slide toggle:', () => {
  it('should render with default props', () => {
    expect(renderSlideToggle().toJSON()).toMatchSnapshot();
  });

  it('should render a checked state', () => {
    expect(renderSlideToggle({ checked: true }).toJSON()).toMatchSnapshot();
  });

  it('should render with the label on the left', () => {
    expect(renderSlideToggle({ position: 'left' }).toJSON()).toMatchSnapshot();
  });

  it('should render with the label hidden', () => {
    expect(renderSlideToggle({ hideLabel: true }).toJSON()).toMatchSnapshot();
  });

  it('should call a passed in onToggle function when the checkbox fires a change event', () => {
    const onToggle = jest.fn();
    const slideToggle = renderIntoDocument(
      <SlideToggle
        id="testToggle"
        label="Test toggle"
        onToggle={onToggle}
      />
    );
    const input = findRenderedDOMComponentWithTag(slideToggle, 'input');
    Simulate.change(input);
    expect(onToggle).toHaveBeenCalled();
  });
});
