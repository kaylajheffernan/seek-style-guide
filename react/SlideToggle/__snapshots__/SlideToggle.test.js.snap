// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slide toggle: should render a checked state 1`] = `
<label
  className="root"
  htmlFor="testToggle"
>
  <div
    className="switch"
  >
    <input
      aria-label="Test toggle"
      checked={true}
      className="input"
      id="testToggle"
      onChange={[Function]}
      type="checkbox"
    />
    <div
      className="slider"
    >
      <div
        className="slideButton"
      >
        <span
          className="root icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": "mock svg",
            }
          }
        />
      </div>
    </div>
  </div>
  <span
    className="label labelRight"
  >
    Test toggle
  </span>
</label>
`;

exports[`Slide toggle: should render child elements 1`] = `
<label
  className="root"
  htmlFor="testToggle"
>
  <div
    className="switch"
  >
    <input
      aria-label="Test toggle"
      checked={undefined}
      className="input"
      id="testToggle"
      onChange={[Function]}
      type="checkbox"
    />
    <div
      className="slider"
    >
      <div
        className="slideButton"
      >
        <span
          className="root icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": "mock svg",
            }
          }
        />
      </div>
    </div>
  </div>
  <span>
    child
  </span>
</label>
`;

exports[`Slide toggle: should render child elements in the left position 1`] = `
<label
  className="root"
  htmlFor="testToggle"
>
  <span>
    child
  </span>
  <div
    className="switch"
  >
    <input
      aria-label="Test toggle"
      checked={undefined}
      className="input"
      id="testToggle"
      onChange={[Function]}
      type="checkbox"
    />
    <div
      className="slider"
    >
      <div
        className="slideButton"
      >
        <span
          className="root icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": "mock svg",
            }
          }
        />
      </div>
    </div>
  </div>
</label>
`;

exports[`Slide toggle: should render with default props 1`] = `
<label
  className="root"
  htmlFor="testToggle"
>
  <div
    className="switch"
  >
    <input
      aria-label="Test toggle"
      checked={undefined}
      className="input"
      id="testToggle"
      onChange={[Function]}
      type="checkbox"
    />
    <div
      className="slider"
    >
      <div
        className="slideButton"
      >
        <span
          className="root icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": "mock svg",
            }
          }
        />
      </div>
    </div>
  </div>
  <span
    className="label labelRight"
  >
    Test toggle
  </span>
</label>
`;

exports[`Slide toggle: should render with the label hidden 1`] = `
<label
  className="root"
  htmlFor="testToggle"
>
  <div
    className="switch"
  >
    <input
      aria-label="Test toggle"
      checked={undefined}
      className="input"
      id="testToggle"
      onChange={[Function]}
      type="checkbox"
    />
    <div
      className="slider"
    >
      <div
        className="slideButton"
      >
        <span
          className="root icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": "mock svg",
            }
          }
        />
      </div>
    </div>
  </div>
</label>
`;

exports[`Slide toggle: should render with the label on the left 1`] = `
<label
  className="root"
  htmlFor="testToggle"
>
  <span
    className="label labelLeft"
  >
    Test toggle
  </span>
  <div
    className="switch"
  >
    <input
      aria-label="Test toggle"
      checked={undefined}
      className="input"
      id="testToggle"
      onChange={[Function]}
      type="checkbox"
    />
    <div
      className="slider"
    >
      <div
        className="slideButton"
      >
        <span
          className="root icon"
          dangerouslySetInnerHTML={
            Object {
              "__html": "mock svg",
            }
          }
        />
      </div>
    </div>
  </div>
</label>
`;
